File: winter2009/comp2150/a5/q2/lib/cashier_departure_event.rb
# file: lib/cashier_departure_event.rb

#Models the departure of a cashier from a store simulation
class CashierDepartureEvent < DepartureEvent

  def initialize(time, cashier)
    super(time, cashier)
    @cashier = cashier
  end

  # Event action. Remove self from the checkout, and record the time.
  def run
    SoopaStore.checkout.cashier = nil
    @cashier.checkout           = nil
    @cashier.departure_time = SoopaStore.current_time
    puts "Time #{SoopaStore.current_time}  : Cashier #{@cashier.eid} "\
    "leaves checkout (worked #{@cashier.time_in_store} time units)"
  end
  
end

File: winter2009/comp2150/a3/q1/CashierDepartureEvent.java
/**
 * Models the departure of an employee in a simulation
 */
class CashierDepartureEvent extends DepartureEvent {

  /**
   * Construct a new CashierDepartureEvent
   *
   * @param time time at which the event occurs
   * @param employee employee leaving
   * @param checkout checkout employee is leaving from
   * @externals Cashier, Checkout
   */
  public CashierDepartureEvent(int time, Cashier employee, Checkout checkout) {
    super(time, employee, checkout);
  }
  
  /**
   * method called when event "happens". Remove self from Checkout.
   * @externals Cashier
   */
  public void action() {
    Cashier employee = (Cashier) this.getActor();
    checkout.setCashier(null);
    System.out.printf("Time %d: Cashier %d leaves checkout %d (worked %d time units)\n",
                      time,
                      employee.getId(),
                      checkout.getIndex(),
                      time-employee.getArrival());
  }
  
}